### ISSUE TITLE
Bug while harvesting and remote viewing

### ISSUE THREAD

**Issue by cdeath-zz** [sp00ktober/DSP_Mod#8]
I'm really good at breaking the game apparently.  Sorry I didn't get a chance to help check that other huge Icarus bug, I haven't seen huge Icarus so I think it's fixed.  I was playing around with a new game so set to harvesting while I scoped out the planets in the system.  I get this error and Icarus starts walking across the remote planet.  My guess is it's trying to "keep harvesting" the minerals I'm on.

```
Loading [Jabbah III] factory
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 0 complete
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

vege count = 20001
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 1 complete : vegetation model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

vein count = 601
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 2 complete : vein model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 3 complete : prebuild model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 4 complete : entity model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 5 complete : create rendering batch and sync buffers, time cost: 0.221 ms
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
```

Reproduce by entering remote planet view while harvesting any resource.  Tested 4 times with and without other mods and had 100% chance of happening.

----- edit

I also see this error in the logs, not sure if this is related to the other, as if they both happen, the one above is the one I see in game (see index out of range error at end)  I edit to show both errors in context maybe that will help

```
Loading [Jabbah IV] factory
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 0 complete
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

refresh music playlist
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

vege count = 0
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 1 complete : vegetation model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

vein count = 577
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 2 complete : vein model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 3 complete : prebuild model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 4 complete : entity model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

step 5 complete : create rendering batch and sync buffers, time cost: 0.219 ms
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

NullReferenceException: Object reference not set to an instance of an object
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0 
 
(Filename:  Line: -1)

refresh music playlist
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

IndexOutOfRangeException: Array index is out of range.
  at PlayerOrder.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at Player.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameData.GameTick (Int64 time) [0x00000] in <filename unknown>:0 
  at GameMain.FixedUpdate () [0x00000] in <filename unknown>:0
 
(Filename:  Line: -1)
```

**Comment by sp00ktober** [comment]
Should be fixed by 58f0f2a
thanks for telling me all these bugs :)

Its not a perfect solution but better than having a big red error in the corner.
@cdeath-zz if you could tell me if this fixed it for you that would be great

**Comment by cdeath-zz** [comment]
The first error seems to be fixed, but while in remote view, Icarus still tries to "get to" the resources and then errors out when it gets to the location.  Now I get a flying Icarus when returning to the main planet, and Icarus has teleported into the sky.  I get the array error now in game.

```
Loading [Vouskiaho - II] factory
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 0 complete
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

vege count = 0
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 1 complete : vegetation model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

vein count = 388
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 2 complete : vein model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 3 complete : prebuild model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 4 complete : entity model & physics
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

step 5 complete : create rendering batch and sync buffers, time cost: 0.251 ms
 
(Filename: D:\unity\Runtime/Export/Debug.bindings.h Line: 45)

IndexOutOfRangeException: Array index is out of range.
at (wrapper dynamic-method) PlayerAction_Mine.DMD<PlayerAction_Mine..GameTick> (PlayerAction_Mine,long) <0x00332>
at PlayerController.GameTick (long) <0x0013f>
at Player.GameTick (long) <0x0006c>
at (wrapper dynamic-method) GameData.DMD<GameData..GameTick> (GameData,long) <0x00110>
at GameMain.FixedUpdate () <0x00262>
```

Further testing has resulted in weird effects like mining random resources on the remote planet or giving the above crash.  This error is not 100% reproduceable.  My guess is Icarus is trying to still mine the resource when planet view is changed.  Is there a way to cancel the mining action when going into remote view?  Fun things happening attached.
![20210311121751_1](https://user-images.githubusercontent.com/4262380/110828823-71f32500-8265-11eb-98c6-163bc0fa26ff.jpg)
![20210311122113_1](https://user-images.githubusercontent.com/4262380/110828842-77e90600-8265-11eb-8033-e3f0194bcd4a.jpg)
![20210311122137_1](https://user-images.githubusercontent.com/4262380/110828853-7ae3f680-8265-11eb-8b16-480f730c32d7.jpg)

**Comment by sp00ktober** [comment]
Ye there is a way but its probably a more complex fix.
Thanks for the extensive testing!
I will work on a fix that does not fall apart that fast :)

**Comment by cdeath-zz** [comment]
Oh I retested without all the other mods and same results.  Icarus tries to walk to some point on the remote planet.  Also this crash when remote view gas giant while harvesting.

```
NullReferenceException: Object reference not set to an instance of an object
  at PlayerFootsteps.PlayFootstepEffect (Boolean lr, Single biomo, Boolean water) [0x00000] in <filename unknown>:0 
  at PlayerFootsteps.LateUpdate () [0x00000] in <filename unknown>:0 
```

The weird "in the sky" destinations is from the mod "Galactic scale" that makes planets other sizes, after this is fixed though, it won't matter since Icarus isn't supposed to be moving.

-edit-
I did notice that you seem to stop the harvesting action because I try to harvest the pod in a new game, then went to remote view, the pod progress is restarted when I return.  So that part seems to be working, but somehow it's getting "reset?" when you go into remote view and making it try to harvest something that isn't there.

**Comment by sp00ktober** [comment]
I did not stop the harvesting command, just prevented the other crash when the game wanted to get information about vegetation from a different planet.
I did that by letting the game access the data from the planet you „really“ are on when doing remote planet view, but as you noticed there are some other things to consider.
The only other thing i did was resetting the mecha core energy to what it was before entering remote planet view if you were currently mining something or walking.

**Comment by sp00ktober** [comment]
[DSP_RenderDistance.zip](https://github.com/sp00ktober/DSP_Mod/files/6134943/DSP_RenderDistance.zip)
@cdeath-zz could you maybe try if this fixed all the issues? the code is already pushed with c2b9549da0f40d9e74f2589180788ac117916dc6

**Comment by cdeath-zz** [comment]
Looks good, I tried to reproduce all the other errors and wasn't able to get any errors.  Icarus doesn't try to move when on remote  planets.  There is a minor visual bug if Icarus is moving to something and you go to remote view, the path disappears, however it does not cause any problems and Icarus continues on this path (line is gone) and does whatever action you had told it to do.

--edit
For good measure I tested with Galactic Scale to see if the different sized planets cause problems and no issues were discovered.
